File: snippets/screens/swiftui/survey_screen/1/survey_screen.swift
import SwiftUI

struct SurveyScreen: View {
    @State private var currentQuestionIndex = 0
    @State private var responses: [String: Any] = [:]
    @State private var isCompleted = false
    @State private var showResults = false

    private let questions = SurveyQuestion.sampleQuestions

    var body: some View {
        NavigationView {
            if isCompleted {
                surveyCompletionView
            } else {
                surveyQuestionView
            }
        }
    }

    private var surveyQuestionView: some View {
        VStack {
            // Progress indicator
            ProgressView(value: Double(currentQuestionIndex + 1), total: Double(questions.count))
                .progressViewStyle(LinearProgressViewStyle())
                .padding()

            HStack {
                Text("Question \(currentQuestionIndex + 1) of \(questions.count)")
                    .font(.caption)
                    .foregroundColor(.gray)
                Spacer()
            }
            .padding(.horizontal)

            ScrollView {
                VStack(alignment: .leading, spacing: 20) {
                    // Question category badge
                    HStack {
                        Text(questions[currentQuestionIndex].category)
                            .font(.caption)
                            .padding(.horizontal, 12)
                            .padding(.vertical, 6)
                            .background(Color.blue.opacity(0.2))
                            .cornerRadius(12)
                            .foregroundColor(.blue)
                        Spacer()
                    }

                    // Question title
                    Text(questions[currentQuestionIndex].title)
                        .font(.title2)
                        .fontWeight(.bold)
                        .multilineTextAlignment(.leading)

                    // Question text
                    Text(questions[currentQuestionIndex].question)
                        .font(.body)
                        .foregroundColor(.secondary)

                    if questions[currentQuestionIndex].isRequired {
                        Text("* Required")
                            .font(.caption)
                            .foregroundColor(.red)
                    }

                    // Question content based on type
                    questionContentView(for: questions[currentQuestionIndex])
                }
                .padding()
            }

            // Navigation buttons
            HStack {
                if currentQuestionIndex > 0 {
                    Button("Previous") {
                        withAnimation {
                            currentQuestionIndex -= 1
                        }
                    }
                    .buttonStyle(SecondaryButtonStyle())
                }

                Spacer()

                Button(currentQuestionIndex == questions.count - 1 ? "Finish Survey" : "Next") {
                    if currentQuestionIndex == questions.count - 1 {
                        completesurvey()
                    } else {
                        withAnimation {
                            currentQuestionIndex += 1
                        }
                    }
                }
                .disabled(!isCurrentQuestionAnswered())
                .buttonStyle(PrimaryButtonStyle())
            }
            .padding()
        }
        .navigationTitle("Survey")
        .navigationBarTitleDisplayMode(.inline)
    }

    private var surveyCompletionView: some View {
        VStack(spacing: 30) {
            Image(systemName: "checkmark.circle.fill")
                .font(.system(size: 80))
                .foregroundColor(.green)

            VStack(spacing: 16) {
                Text("Thank You!")
                    .font(.largeTitle)
                    .fontWeight(.bold)

                Text("Your responses have been submitted successfully. We appreciate your time and valuable feedback.")
                    .font(.body)
                    .multilineTextAlignment(.center)
                    .foregroundColor(.secondary)
            }

            VStack(spacing: 16) {
                Text("Survey Summary")
                    .font(.headline)

                VStack(alignment: .leading, spacing: 8) {
                    HStack {
                        Text("Questions answered:")
                        Spacer()
                        Text("\(responses.count) of \(questions.count)")
                            .fontWeight(.semibold)
                    }

                    HStack {
                        Text("Completion rate:")
                        Spacer()
                        Text("\(Int(Double(responses.count) / Double(questions.count) * 100))%")
                            .fontWeight(.semibold)
                            .foregroundColor(.green)
                    }
                }
                .padding()
                .background(Color.gray.opacity(0.1))
                .cornerRadius(10)
            }

            Button("Close") {
                // Handle close action
            }
            .buttonStyle(PrimaryButtonStyle())
        }
        .padding()
        .navigationTitle("Survey Complete")
        .navigationBarTitleDisplayMode(.inline)
    }

    @ViewBuilder
    private func questionContentView(for question: SurveyQuestion) -> some View {
        switch question.type {
        case .singleChoice, .multipleChoice:
            choiceQuestionView(for: question)
        case .checkbox:
            checkboxQuestionView(for: question)
        case .scale:
            scaleQuestionView(for: question)
        case .text:
            textQuestionView(for: question)
        case .yesNo:
            yesNoQuestionView(for: question)
        }
    }

    @ViewBuilder
    private func choiceQuestionView(for question: SurveyQuestion) -> some View {
        VStack(alignment: .leading, spacing: 12) {
            ForEach(question.options ?? [], id: \.self) { option in
                Button(action: {
                    responses[question.id] = option
                }) {
                    HStack {
                        Image(systemName: responses[question.id] as? String == option ? "largecircle.fill.circle" : "circle")
                            .foregroundColor(.blue)
                        Text(option)
                            .foregroundColor(.primary)
                        Spacer()
                    }
                    .padding()
                    .background(
                        RoundedRectangle(cornerRadius: 10)
                            .stroke(responses[question.id] as? String == option ? Color.blue : Color.gray.opacity(0.3), lineWidth: 2)
                    )
                }
                .buttonStyle(PlainButtonStyle())
            }
        }
    }

    @ViewBuilder
    private func checkboxQuestionView(for question: SurveyQuestion) -> some View {
        VStack(alignment: .leading, spacing: 12) {
            ForEach(question.options ?? [], id: \.self) { option in
                Button(action: {
                    var selectedOptions = responses[question.id] as? [String] ?? []
                    if selectedOptions.contains(option) {
                        selectedOptions.removeAll { $0 == option }
                    } else {
                        selectedOptions.append(option)
                    }
                    responses[question.id] = selectedOptions
                }) {
                    HStack {
                        let selectedOptions = responses[question.id] as? [String] ?? []
                        Image(systemName: selectedOptions.contains(option) ? "checkmark.square.fill" : "square")
                            .foregroundColor(.blue)
                        Text(option)
                            .foregroundColor(.primary)
                        Spacer()
                    }
                    .padding()
                    .background(
                        RoundedRectangle(cornerRadius: 10)
                            .stroke((responses[question.id] as? [String])?.contains(option) == true ? Color.blue : Color.gray.opacity(0.3), lineWidth: 2)
                    )
                }
                .buttonStyle(PlainButtonStyle())
            }
        }
    }

    @ViewBuilder
    private func scaleQuestionView(for question: SurveyQuestion) -> some View {
        VStack(spacing: 16) {
            let currentValue = responses[question.id] as? Double ?? Double(question.scaleMin ?? 1)

            Text("Rating: \(Int(currentValue))/\(question.scaleMax ?? 10)")
                .font(.headline)

            Slider(
                value: Binding(
                    get: { responses[question.id] as? Double ?? Double(question.scaleMin ?? 1) },
                    set: { responses[question.id] = $0 }
                ),
                in: Double(question.scaleMin ?? 1)...Double(question.scaleMax ?? 10),
                step: 1
            )
            .accentColor(.blue)

            HStack {
                Text("\(question.scaleMin ?? 1)")
                    .font(.caption)
                    .foregroundColor(.gray)
                Spacer()
                Text("\(question.scaleMax ?? 10)")
                    .font(.caption)
                    .foregroundColor(.gray)
            }
        }
        .padding()
        .background(Color.gray.opacity(0.05))
        .cornerRadius(10)
    }

    @ViewBuilder
    private func textQuestionView(for question: SurveyQuestion) -> some View {
        TextEditor(text: Binding(
            get: { responses[question.id] as? String ?? "" },
            set: { responses[question.id] = $0 }
        ))
        .frame(minHeight: 100)
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(Color.gray.opacity(0.3), lineWidth: 1)
        )
    }

    @ViewBuilder
    private func yesNoQuestionView(for question: SurveyQuestion) -> some View {
        HStack(spacing: 20) {
            Button(action: {
                responses[question.id] = true
            }) {
                HStack {
                    Image(systemName: responses[question.id] as? Bool == true ? "largecircle.fill.circle" : "circle")
                        .foregroundColor(.blue)
                    Text("Yes")
                        .foregroundColor(.primary)
                }
                .frame(maxWidth: .infinity)
                .padding()
                .background(
                    RoundedRectangle(cornerRadius: 10)
                        .stroke(responses[question.id] as? Bool == true ? Color.blue : Color.gray.opacity(0.3), lineWidth: 2)
                )
            }
            .buttonStyle(PlainButtonStyle())

            Button(action: {
                responses[question.id] = false
            }) {
                HStack {
                    Image(systemName: responses[question.id] as? Bool == false ? "largecircle.fill.circle" : "circle")
                        .foregroundColor(.blue)
                    Text("No")
                        .foregroundColor(.primary)
                }
                .frame(maxWidth: .infinity)
                .padding()
                .background(
                    RoundedRectangle(cornerRadius: 10)
                        .stroke(responses[question.id] as? Bool == false ? Color.blue : Color.gray.opacity(0.3), lineWidth: 2)
                )
            }
            .buttonStyle(PlainButtonStyle())
        }
    }

    private func isCurrentQuestionAnswered() -> Bool {
        let question = questions[currentQuestionIndex]
        let response = responses[question.id]

        if !question.isRequired { return true }

        switch question.type {
        case .singleChoice, .multipleChoice, .yesNo:
            return response != nil
        case .checkbox:
            return (response as? [String])?.isEmpty == false
        case .scale:
            return response != nil
        case .text:
            return !(response as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }

    private func completeSurvey() {
        withAnimation {
            isCompleted = true
        }
    }
}

struct SurveyQuestion {
    let id: String
    let title: String
    let type: QuestionType
    let question: String
    let isRequired: Bool
    let options: [String]?
    let scaleMin: Int?
    let scaleMax: Int?
    let placeholder: String?
    let category: String

    enum QuestionType {
        case singleChoice
        case multipleChoice
        case checkbox
        case scale
        case text
        case yesNo
    }

    static let sampleQuestions = [
        SurveyQuestion(
            id: "demographics",
            title: "Tell us about yourself",
            type: .singleChoice,
            question: "What is your age group?",
            isRequired: true,
            options: ["18-24", "25-34", "35-44", "45-54", "55-64", "65+"],
            scaleMin: nil,
            scaleMax: nil,
            placeholder: nil,
            category: "Demographics"
        ),
        SurveyQuestion(
            id: "usage_frequency",
            title: "Usage Patterns",
            type: .singleChoice,
            question: "How often do you use our app?",
            isRequired: true,
            options: ["Daily", "Weekly", "Monthly", "Rarely", "First time"],
            scaleMin: nil,
            scaleMax: nil,
            placeholder: nil,
            category: "Usage"
        ),
        SurveyQuestion(
            id: "satisfaction_rating",
            title: "Satisfaction Rating",
            type: .scale,
            question: "How satisfied are you with our service?",
            isRequired: true,
            options: nil,
            scaleMin: 1,
            scaleMax: 10,
            placeholder: nil,
            category: "Feedback"
        ),
        SurveyQuestion(
            id: "features_used",
            title: "Feature Usage",
            type: .checkbox,
            question: "Which features do you use most? (Select all that apply)",
            isRequired: false,
            options: ["Dashboard", "Reports", "Settings", "Search", "Notifications", "Export Data"],
            scaleMin: nil,
            scaleMax: nil,
            placeholder: nil,
            category: "Features"
        ),
        SurveyQuestion(
            id: "improvement_suggestions",
            title: "Your Feedback",
            type: .text,
            question: "What improvements would you like to see?",
            isRequired: false,
            options: nil,
            scaleMin: nil,
            scaleMax: nil,
            placeholder: "Share your thoughts and suggestions...",
            category: "Feedback"
        ),
        SurveyQuestion(
            id: "recommendation",
            title: "Final Question",
            type: .yesNo,
            question: "Would you recommend our app to others?",
            isRequired: true,
            options: nil,
            scaleMin: nil,
            scaleMax: nil,
            placeholder: nil,
            category: "Recommendation"
        )
    ]
}

struct SecondaryButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .padding()
            .background(Color.gray.opacity(0.2))
            .foregroundColor(.primary)
            .cornerRadius(10)
            .scaleEffect(configuration.isPressed ? 0.95 : 1.0)
    }
}

struct SurveyScreen_Previews: PreviewProvider {
    static var previews: some View {
        SurveyScreen()
    }
}

File: snippets/screens/flutter/survey_screen/1/survey_screen.dart
import 'package:flutter/material.dart';

class SurveyScreen extends StatefulWidget {
  const SurveyScreen({Key? key}) : super(key: key);

  @override
  State<SurveyScreen> createState() => _SurveyScreenState();
}

class _SurveyScreenState extends State<SurveyScreen> {
  final PageController _pageController = PageController();
  int _currentQuestionIndex = 0;
  bool _isSubmitting = false;

  final Map<String, dynamic> _responses = {};

  final List<SurveyQuestion> _questions = [
    SurveyQuestion(
      id: 'demographics',
      title: 'Tell us about yourself',
      type: QuestionType.multipleChoice,
      question: 'What is your age group?',
      isRequired: true,
      options: ['18-24', '25-34', '35-44', '45-54', '55-64', '65+'],
    ),
    SurveyQuestion(
      id: 'usage_frequency',
      title: 'Usage Patterns',
      type: QuestionType.singleChoice,
      question: 'How often do you use our app?',
      isRequired: true,
      options: ['Daily', 'Weekly', 'Monthly', 'Rarely', 'First time'],
    ),
    SurveyQuestion(
      id: 'satisfaction_rating',
      title: 'Satisfaction Rating',
      type: QuestionType.scale,
      question: 'How satisfied are you with our service?',
      isRequired: true,
      scaleMin: 1,
      scaleMax: 10,
    ),
    SurveyQuestion(
      id: 'features_used',
      title: 'Feature Usage',
      type: QuestionType.checkbox,
      question: 'Which features do you use most? (Select all that apply)',
      isRequired: false,
      options: [
        'Dashboard',
        'Reports',
        'Settings',
        'Search',
        'Notifications',
        'Export Data',
      ],
    ),
    SurveyQuestion(
      id: 'improvement_suggestions',
      title: 'Your Feedback',
      type: QuestionType.text,
      question: 'What improvements would you like to see?',
      isRequired: false,
      placeholder: 'Share your thoughts and suggestions...',
    ),
    SurveyQuestion(
      id: 'recommendation',
      title: 'Final Question',
      type: QuestionType.yesNo,
      question: 'Would you recommend our app to others?',
      isRequired: true,
    ),
  ];

  @override
  void dispose() {
    _pageController.dispose();
    super.dispose();
  }

  bool _isCurrentQuestionAnswered() {
    final question = _questions[_currentQuestionIndex];
    final response = _responses[question.id];

    if (!question.isRequired) return true;

    switch (question.type) {
      case QuestionType.singleChoice:
      case QuestionType.multipleChoice:
      case QuestionType.yesNo:
        return response != null;
      case QuestionType.checkbox:
        return response is List && response.isNotEmpty;
      case QuestionType.scale:
        return response != null;
      case QuestionType.text:
        return response != null && response.toString().trim().isNotEmpty;
    }
  }

  void _nextQuestion() {
    if (_currentQuestionIndex < _questions.length - 1) {
      setState(() => _currentQuestionIndex++);
      _pageController.nextPage(
        duration: const Duration(milliseconds: 300),
        curve: Curves.easeInOut,
      );
    }
  }

  void _previousQuestion() {
    if (_currentQuestionIndex > 0) {
      setState(() => _currentQuestionIndex--);
      _pageController.previousPage(
        duration: const Duration(milliseconds: 300),
        curve: Curves.easeInOut,
      );
    }
  }

  void _submitSurvey() async {
    setState(() => _isSubmitting = true);

    // Simulate API call
    await Future.delayed(const Duration(seconds: 2));

    setState(() => _isSubmitting = false);

    if (mounted) {
      showDialog(
        context: context,
        barrierDismissible: false,
        builder: (context) => AlertDialog(
          title: const Text('Thank You!'),
          content: const Text('Your survey responses have been submitted successfully.'),
          actions: [
            TextButton(
              onPressed: () {
                Navigator.of(context).pop();
                Navigator.of(context).pop();
              },
              child: const Text('Close'),
            ),
          ],
        ),
      );
    }
  }

  Widget _buildQuestionWidget(SurveyQuestion question) {
    switch (question.type) {
      case QuestionType.singleChoice:
      case QuestionType.multipleChoice:
        return _buildChoiceQuestion(question);
      case QuestionType.checkbox:
        return _buildCheckboxQuestion(question);
      case QuestionType.scale:
        return _buildScaleQuestion(question);
      case QuestionType.text:
        return _buildTextQuestion(question);
      case QuestionType.yesNo:
        return _buildYesNoQuestion(question);
    }
  }

  Widget _buildChoiceQuestion(SurveyQuestion question) {
    return Column(
      children: question.options!.map((option) {
        return RadioListTile<String>(
          title: Text(option),
          value: option,
          groupValue: _responses[question.id],
          onChanged: (value) {
            setState(() => _responses[question.id] = value);
          },
        );
      }).toList(),
    );
  }

  Widget _buildCheckboxQuestion(SurveyQuestion question) {
    List<String> selectedOptions = _responses[question.id] ?? [];

    return Column(
      children: question.options!.map((option) {
        return CheckboxListTile(
          title: Text(option),
          value: selectedOptions.contains(option),
          onChanged: (checked) {
            setState(() {
              if (checked == true) {
                selectedOptions.add(option);
              } else {
                selectedOptions.remove(option);
              }
              _responses[question.id] = selectedOptions;
            });
          },
        );
      }).toList(),
    );
  }

  Widget _buildScaleQuestion(SurveyQuestion question) {
    double currentValue = (_responses[question.id] ?? question.scaleMin!).toDouble();

    return Column(
      children: [
        Text(
          'Rating: ${currentValue.toInt()}/${question.scaleMax}',
          style: Theme.of(context).textTheme.titleLarge,
        ),
        const SizedBox(height: 16),
        Slider(
          value: currentValue,
          min: question.scaleMin!.toDouble(),
          max: question.scaleMax!.toDouble(),
          divisions: question.scaleMax! - question.scaleMin!,
          label: currentValue.toInt().toString(),
          onChanged: (value) {
            setState(() => _responses[question.id] = value.toInt());
          },
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text('${question.scaleMin}'),
            Text('${question.scaleMax}'),
          ],
        ),
      ],
    );
  }

  Widget _buildTextQuestion(SurveyQuestion question) {
    return TextFormField(
      initialValue: _responses[question.id] ?? '',
      decoration: InputDecoration(
        hintText: question.placeholder,
        border: const OutlineInputBorder(),
        alignLabelWithHint: true,
      ),
      maxLines: 4,
      maxLength: 500,
      onChanged: (value) => _responses[question.id] = value,
    );
  }

  Widget _buildYesNoQuestion(SurveyQuestion question) {
    return Row(
      children: [
        Expanded(
          child: RadioListTile<bool>(
            title: const Text('Yes'),
            value: true,
            groupValue: _responses[question.id],
            onChanged: (value) => setState(() => _responses[question.id] = value),
          ),
        ),
        Expanded(
          child: RadioListTile<bool>(
            title: const Text('No'),
            value: false,
            groupValue: _responses[question.id],
            onChanged: (value) => setState(() => _responses[question.id] = value),
          ),
        ),
      ],
    );
  }

  @override
  Widget build(BuildContext context) {
    final question = _questions[_currentQuestionIndex];
    final progress = (_currentQuestionIndex + 1) / _questions.length;

    return Scaffold(
      appBar: AppBar(
        title: const Text('Survey'),
        elevation: 0,
        bottom: PreferredSize(
          preferredSize: const Size.fromHeight(4),
          child: LinearProgressIndicator(value: progress),
        ),
      ),
      body: PageView.builder(
        controller: _pageController,
        physics: const NeverScrollableScrollPhysics(),
        itemCount: _questions.length,
        itemBuilder: (context, index) {
          final question = _questions[index];
          return SingleChildScrollView(
            padding: const EdgeInsets.all(16.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Card(
                  child: Padding(
                    padding: const EdgeInsets.all(16.0),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Row(
                          children: [
                            CircleAvatar(
                              child: Text('${index + 1}'),
                              backgroundColor: Theme.of(context).primaryColor,
                              foregroundColor: Colors.white,
                            ),
                            const SizedBox(width: 12),
                            Expanded(
                              child: Column(
                                crossAxisAlignment: CrossAxisAlignment.start,
                                children: [
                                  Text(
                                    question.title,
                                    style: Theme.of(context).textTheme.titleMedium,
                                  ),
                                  const SizedBox(height: 4),
                                  Text(
                                    '${index + 1} of ${_questions.length}',
                                    style: Theme.of(context).textTheme.bodySmall,
                                  ),
                                ],
                              ),
                            ),
                          ],
                        ),
                        const SizedBox(height: 16),
                        Text(
                          question.question,
                          style: Theme.of(context).textTheme.titleLarge,
                        ),
                        if (question.isRequired)
                          const Text(
                            '* Required',
                            style: TextStyle(color: Colors.red, fontSize: 12),
                          ),
                        const SizedBox(height: 20),
                        _buildQuestionWidget(question),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          );
        },
      ),
      bottomNavigationBar: Container(
        padding: const EdgeInsets.all(16.0),
        child: Row(
          children: [
            if (_currentQuestionIndex > 0)
              Expanded(
                child: OutlinedButton(
                  onPressed: _previousQuestion,
                  child: const Text('Previous'),
                ),
              ),
            if (_currentQuestionIndex > 0) const SizedBox(width: 16),
            Expanded(
              flex: 2,
              child: ElevatedButton(
                onPressed: _currentQuestionIndex == _questions.length - 1
                    ? (_isCurrentQuestionAnswered() ? _submitSurvey : null)
                    : (_isCurrentQuestionAnswered() ? _nextQuestion : null),
                child: _isSubmitting
                    ? const SizedBox(
                        height: 20,
                        width: 20,
                        child: CircularProgressIndicator(strokeWidth: 2),
                      )
                    : Text(_currentQuestionIndex == _questions.length - 1
                        ? 'Submit Survey'
                        : 'Next'),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class SurveyQuestion {
  final String id;
  final String title;
  final QuestionType type;
  final String question;
  final bool isRequired;
  final List<String>? options;
  final int? scaleMin;
  final int? scaleMax;
  final String? placeholder;

  SurveyQuestion({
    required this.id,
    required this.title,
    required this.type,
    required this.question,
    this.isRequired = false,
    this.options,
    this.scaleMin,
    this.scaleMax,
    this.placeholder,
  });
}

enum QuestionType {
  singleChoice,
  multipleChoice,
  checkbox,
  scale,
  text,
  yesNo,
}